File: tests/wasm/test_rsx.js
/**
 * WASM binding tests for RSX indicator.
 * These tests mirror the Rust unit tests to ensure WASM bindings work correctly.
 */
import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';
import { 
    loadTestData, 
    assertArrayClose, 
    assertClose,
    isNaN,
    assertAllNaN,
    assertNoNaN,
    EXPECTED_OUTPUTS 
} from './test_utils.js';
import { compareWithRust } from './rust-comparison.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let wasm;
let testData;

test.before(async () => {
    
    try {
        const wasmPath = path.join(__dirname, '../../pkg/my_project.js');
        const importPath = process.platform === 'win32' 
            ? 'file:///' + wasmPath.replace(/\\/g, '/')
            : wasmPath;
        wasm = await import(importPath);
        
    } catch (error) {
        console.error('Failed to load WASM module. Run "wasm-pack build --features wasm --target nodejs" first');
        throw error;
    }
    
    testData = loadTestData();
});

test('RSX partial params', () => {
    
    const close = new Float64Array(testData.close);
    
    const result = wasm.rsx_js(close, 14);
    assert.strictEqual(result.length, close.length);
});

test('RSX accuracy', async () => {
    
    const close = new Float64Array(testData.close);
    const expected = EXPECTED_OUTPUTS.rsx;
    
    const result = wasm.rsx_js(
        close,
        expected.default_params.period
    );
    
    assert.strictEqual(result.length, close.length);
    
    
    const last5 = result.slice(-5);
    assertArrayClose(
        last5,
        expected.last_5_values,
        0.1,  
        "RSX last 5 values mismatch"
    );
    
    
    await compareWithRust('rsx', result, 'close', expected.default_params);
});

test('RSX default candles', () => {
    
    const close = new Float64Array(testData.close);
    
    const result = wasm.rsx_js(close, 14);
    assert.strictEqual(result.length, close.length);
});

test('RSX zero period', () => {
    
    const inputData = new Float64Array([10.0, 20.0, 30.0]);
    
    assert.throws(() => {
        wasm.rsx_js(inputData, 0);
    }, /Invalid period/);
});

test('RSX period exceeds length', () => {
    
    const dataSmall = new Float64Array([10.0, 20.0, 30.0]);
    
    assert.throws(() => {
        wasm.rsx_js(dataSmall, 10);
    }, /Invalid period/);
});

test('RSX all NaN', () => {
    
    const allNan = new Float64Array(10).fill(NaN);
    
    assert.throws(() => {
        wasm.rsx_js(allNan, 3);
    }, /All values are NaN/);
});

test('RSX NaN handling', () => {
    
    const close = new Float64Array(testData.close);
    
    const result = wasm.rsx_js(close, 14);
    assert.strictEqual(result.length, close.length);
    
    
    if (result.length > 240) {
        for (let i = 240; i < result.length; i++) {
            assert(!isNaN(result[i]), `Unexpected NaN at index ${i}`);
        }
    }
});

test('RSX empty input', () => {
    
    const emptyData = new Float64Array([]);
    
    assert.throws(() => {
        wasm.rsx_js(emptyData, 14);
    }, /All values are NaN|Invalid period|empty/i);
});

test('RSX fast API (rsx_into)', () => {
    
    const close = new Float64Array(testData.close);
    const len = close.length;
    
    
    const outPtr = wasm.rsx_alloc(len);
    const inPtr = wasm.rsx_alloc(len);
    
    try {
        
        const memory = new Float64Array(wasm.__wasm.memory.buffer, inPtr, len);
        memory.set(close);
        
        
        wasm.rsx_into(inPtr, outPtr, len, 14);
        
        
        const result = new Float64Array(wasm.__wasm.memory.buffer, outPtr, len);
        const resultCopy = new Float64Array(result);
        
        
        const safeResult = wasm.rsx_js(close, 14);
        assertArrayClose(resultCopy, safeResult, 1e-10, "Fast API mismatch with safe API");
    } finally {
        
        wasm.rsx_free(inPtr, len);
        wasm.rsx_free(outPtr, len);
    }
});

test('RSX fast API in-place', () => {
    
    const close = new Float64Array(testData.close);
    const len = close.length;
    
    
    const ptr = wasm.rsx_alloc(len);
    
    try {
        
        const memory = new Float64Array(wasm.__wasm.memory.buffer, ptr, len);
        memory.set(close);
        
        
        wasm.rsx_into(ptr, ptr, len, 14);
        
        
        const result = new Float64Array(wasm.__wasm.memory.buffer, ptr, len);
        const resultCopy = new Float64Array(result);
        
        
        const safeResult = wasm.rsx_js(close, 14);
        assertArrayClose(resultCopy, safeResult, 1e-10, "In-place operation mismatch");
    } finally {
        
        wasm.rsx_free(ptr, len);
    }
});

test('RSX memory allocation/deallocation', () => {
    
    const len = 1000;
    
    
    const ptr = wasm.rsx_alloc(len);
    assert(ptr !== 0, "Allocation returned null pointer");
    
    
    const memory = new Float64Array(wasm.__wasm.memory.buffer, ptr, len);
    for (let i = 0; i < len; i++) {
        memory[i] = i * 1.5;
    }
    
    
    assert.strictEqual(memory[0], 0);
    assert.strictEqual(memory[10], 15);
    
    
    wasm.rsx_free(ptr, len);
    
    
    const ptr2 = wasm.rsx_alloc(len);
    assert(ptr2 !== 0, "Re-allocation failed");
    wasm.rsx_free(ptr2, len);
});

test('RSX null pointer handling', () => {
    
    assert.throws(() => {
        wasm.rsx_into(0, 0, 100, 14);
    }, /null pointer/);
});

test('RSX batch API', () => {
    
    const close = new Float64Array(testData.close);
    
    const config = {
        period_range: [10, 20, 5]  
    };
    
    const result = wasm.rsx_batch(close, config);
    
    assert(result.values, "Batch result should have values");
    assert(result.combos, "Batch result should have combos");
    assert.strictEqual(result.rows, 3, "Should have 3 parameter combinations");
    assert.strictEqual(result.cols, close.length, "Columns should match data length");
    assert.strictEqual(result.values.length, result.rows * result.cols, "Values array size mismatch");
    assert.strictEqual(result.combos.length, 3, "Should have 3 parameter combinations");
    
    
    assert.strictEqual(result.combos[0].period, 10);
    assert.strictEqual(result.combos[1].period, 15);
    assert.strictEqual(result.combos[2].period, 20);
});

test('RSX batch fast API', () => {
    
    const close = new Float64Array(testData.close);
    const len = close.length;
    
    
    const periodStart = 10, periodEnd = 20, periodStep = 5;
    const expectedRows = Math.floor((periodEnd - periodStart) / periodStep) + 1; 
    const outputSize = expectedRows * len;
    
    
    const inPtr = wasm.rsx_alloc(len);
    const outPtr = wasm.rsx_alloc(outputSize);
    
    try {
        
        const memory = new Float64Array(wasm.__wasm.memory.buffer, inPtr, len);
        memory.set(close);
        
        
        const rows = wasm.rsx_batch_into(
            inPtr, outPtr, len,
            periodStart, periodEnd, periodStep
        );
        
        assert.strictEqual(rows, expectedRows, "Row count mismatch");
        
        
        const result = new Float64Array(wasm.__wasm.memory.buffer, outPtr, outputSize);
        
        
        assert(isNaN(result[0]), "First value should be NaN");
        
        
        assert(result.length === outputSize, "Output size mismatch");
    } finally {
        
        wasm.rsx_free(inPtr, len);
        wasm.rsx_free(outPtr, outputSize);
    }
});

File: tests/python/test_rsx.py
"""
Python binding tests for RSX indicator.
These tests mirror the Rust unit tests to ensure Python bindings work correctly.
"""
import pytest
import numpy as np
import sys
from pathlib import Path

try:
    import my_project as ta_indicators
except ImportError:
    
    try:
        import my_project as ta_indicators
    except ImportError:
        pytest.skip("Python module not built. Run 'maturin develop --features python' first", allow_module_level=True)

from test_utils import load_test_data, assert_close, EXPECTED_OUTPUTS
from rust_comparison import compare_with_rust


class TestRsx:
    @pytest.fixture(scope='class')
    def test_data(self):
        return load_test_data()
    
    def test_rsx_partial_params(self, test_data):
        """Test RSX with partial parameters (None values) - mirrors check_rsx_partial_params"""
        close = test_data['close']
        
        
        result = ta_indicators.rsx(close, 14)  
        assert len(result) == len(close)
    
    def test_rsx_accuracy(self, test_data):
        """Test RSX matches expected values from Rust tests - mirrors check_rsx_accuracy"""
        close = test_data['close']
        expected = EXPECTED_OUTPUTS['rsx']
        
        result = ta_indicators.rsx(
            close,
            period=expected['default_params']['period']
        )
        
        assert len(result) == len(close)
        
        
        assert_close(
            result[-5:], 
            expected['last_5_values'],
            rtol=1e-1,  
            msg="RSX last 5 values mismatch"
        )
        
        
        compare_with_rust('rsx', result, 'close', expected['default_params'])
    
    def test_rsx_default_candles(self, test_data):
        """Test RSX with default parameters - mirrors check_rsx_default_candles"""
        close = test_data['close']
        
        
        result = ta_indicators.rsx(close, 14)
        assert len(result) == len(close)
    
    def test_rsx_zero_period(self):
        """Test RSX fails with zero period - mirrors check_rsx_zero_period"""
        input_data = np.array([10.0, 20.0, 30.0])
        
        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.rsx(input_data, period=0)
    
    def test_rsx_period_exceeds_length(self):
        """Test RSX fails when period exceeds data length - mirrors check_rsx_period_exceeds_length"""
        data_small = np.array([10.0, 20.0, 30.0])
        
        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.rsx(data_small, period=10)
    
    def test_rsx_very_small_dataset(self):
        """Test RSX fails with insufficient data - mirrors check_rsx_very_small_dataset"""
        single_point = np.array([42.0])
        
        with pytest.raises(ValueError, match="Invalid period|Not enough valid data"):
            ta_indicators.rsx(single_point, period=14)
    
    def test_rsx_reinput(self, test_data):
        """Test RSX applied twice (re-input) - mirrors check_rsx_reinput"""
        close = test_data['close']
        
        
        first_result = ta_indicators.rsx(close, period=14)
        assert len(first_result) == len(close)
        
        
        second_result = ta_indicators.rsx(first_result, period=14)
        assert len(second_result) == len(first_result)
    
    def test_rsx_nan_handling(self, test_data):
        """Test RSX handles NaN values correctly - mirrors check_rsx_nan_handling"""
        close = test_data['close']
        
        result = ta_indicators.rsx(close, period=14)
        assert len(result) == len(close)
        
        
        if len(result) > 50:
            assert not np.any(np.isnan(result[50:])), "Found unexpected NaN after warmup period"
    
    def test_rsx_streaming(self, test_data):
        """Test RSX streaming matches batch calculation - mirrors check_rsx_streaming"""
        close = test_data['close']
        period = 14
        
        
        batch_result = ta_indicators.rsx(close, period=period)
        
        
        stream = ta_indicators.RsxStream(period=period)
        stream_values = []
        
        for price in close:
            result = stream.update(price)
            stream_values.append(result if result is not None else np.nan)
        
        stream_values = np.array(stream_values)
        
        
        assert len(batch_result) == len(stream_values)
        
        
        for i, (b, s) in enumerate(zip(batch_result, stream_values)):
            if np.isnan(b) and np.isnan(s):
                continue
            assert_close(b, s, rtol=1e-9, atol=1e-9, 
                        msg=f"RSX streaming mismatch at index {i}")
    
    def test_rsx_batch(self, test_data):
        """Test RSX batch processing - mirrors check_batch_default_row"""
        close = test_data['close']
        
        result = ta_indicators.rsx_batch(
            close,
            period_range=(14, 100, 1)  
        )
        
        assert 'values' in result
        assert 'periods' in result
        
        
        periods = result['periods']
        default_idx = None
        for i, p in enumerate(periods):
            if p == 14:
                default_idx = i
                break
        
        assert default_idx is not None, "Default period (14) not found in batch results"
        
        
        values_2d = result['values']
        default_row = values_2d[default_idx]
        expected = EXPECTED_OUTPUTS['rsx']['last_5_values']
        
        
        assert_close(
            default_row[-5:],
            expected,
            rtol=1e-1,  
            msg="RSX batch default row mismatch"
        )
    
    def test_rsx_all_nan_input(self):
        """Test RSX with all NaN values"""
        all_nan = np.full(100, np.nan)
        
        with pytest.raises(ValueError, match="All values are NaN"):
            ta_indicators.rsx(all_nan, period=14)
    
    def test_rsx_kernel_parameter(self, test_data):
        """Test RSX with explicit kernel parameter"""
        close = test_data['close']
        
        
        result_scalar = ta_indicators.rsx(close, period=14, kernel="scalar")
        assert len(result_scalar) == len(close)
        
        
        result_auto = ta_indicators.rsx(close, period=14, kernel="auto")
        assert len(result_auto) == len(close)
        
        
        assert_close(result_scalar, result_auto, rtol=1e-9, atol=1e-9)


if __name__ == '__main__':
    pytest.main([__file__, '-v'])